RISDEV-4434 create page for reference documentation no functionality

malte-laukoetter opened the pull request.

malte-laukoetter committed:
Create very basic page layout

RISDEV-4434

malte-laukoetter committed:
Add simple component for selecting a mod

RISDEV-4434

malte-laukoetter committed:
Add simple component for selecting a akn:ref

RISDEV-4434

malte-laukoetter committed:
Add simple component for table of akn:refs

RISDEV-4434

malte-laukoetter committed:
Correctly position save button

RISDEV-4434

malte-laukoetter committed:
Add state for no selected akn:mod element

RISDEV-4434

malte-laukoetter committed:
Improve heading styles

RISDEV-4434

malte-laukoetter committed:
Add Ris prefix to components

RISDEV-4434

malte-laukoetter committed:
Add test for RisModSelectionPanel

RISDEV-4435

malte-laukoetter committed:
Show quoted content instead of full mod

RISDEV-4436

malte-laukoetter committed:
Support creating new akn:ref elements by highlighting text

RISDEV-4436

malte-laukoetter committed:
Add tests for RisRefSelectionPanel

RISDEV-4436

malte-laukoetter committed:
Add selection of ref to RisRefSelectionPanel

RISDEV-4436

malte-laukoetter committed:
Support deleting refs from the table

RISDEV-4437

andreasphil commented on frontend/src/components/references/RisModSelectionPanel.vue:
```diff
@@ -0,0 +1,57 @@
+
+<template>
+  <div class="flex flex-col">
+    <div v-if="normRenderLoading" class="flex justify-center">
```
Suggestion: I know replacing the UI with the spinner is our usual pattern but in this case it makes the UI feel a bit janky when showing a spinner for a split-second after every highlight or removing of a highlight. Some possible alternatives that come to mind:

- Debounce the spinner
- Show the spinner with a semi-transparent overlay on top of the UI instead of replacing it (possibly combined with debouncing)
- Show a small spinner inline next to the heading

andreasphil commented on frontend/src/components/references/RisModSelectionPanel.vue:
```diff
@@ -0,0 +1,57 @@
+
+<style scoped>
+:deep(.akn-mod) {
+  @apply block border border-[#4299F7] bg-[#E7E7E766] px-2;
```
Question: should we add these colors to our Tailwind config rather than hardcoding/arbitrary variants? Or do we know they won't be used anywhere else?

malte-laukoetter replied:
I think we should but they are currently different from the colors already defined for mods in other views so I don't know what to call them. Will talk to design about this ^^

andreasphil commented on frontend/src/components/references/RisModSelectionPanel.vue:
```diff
@@ -0,0 +1,57 @@
+</template>
+
+<style scoped>
+:deep(.akn-mod) {
```
Question: I'm not a huge fan of using `:deep`. I think it's fine for the preview component because of the `v-html`, but I would prefer if we didn't have this all across the app because it makes it harder to understand where styles come from, and harder to change the preview component as it also needs to work with all kinds of unpredictable styling overrides. Maybe we can find a different way to solve this?

andreasphil replied:
Side note: If we break out of the scoping anyways, maybe `style module` or simply `style` would be cleaner?

malte-laukoetter replied:
The previous solution was to have all the different highlighting styles within the preview component itself (and properties to turn them on or off), but that became quite messy with the amount of different highlighting we want to have on different pages. 

The scoped style still (even with the deep) ensures that nothing outside of the component is affected by the style (but other than without the deep child components can be affected). So it still has advantages above a simple `style`. With `style module` I'm unsure how to achieve this kind of styling.

So yes it's not ideal but I couldn't find a better solution.

andreasphil commented on frontend/src/components/references/RisRefEditor.vue:
```diff
@@ -0,0 +1,23 @@
+  <RisTextButton
+    size="small"
+    variant="ghost"
+    label="Löschen"
```
Question: do we need to generate a more dynamic/precise label here? I don't know how this will be logically grouped with the other inputs in this row but I'm not sure simply saying "Löschen" will be enough from the accessibility perspective to make the connection to what is being deleted.

malte-laukoetter replied:
I'll look into this when I implement the other inputs of the component

malte-laukoetter replied:
I have grouped the inputs into sections which have the text content of the akn:refs as names

andreasphil commented on frontend/src/services/xmlService.ts:
```diff
@@ -1,5 +1,5 @@
 /**
- * Convert the text representation of an xml document to a Document
+ * Convert the text representation of a xml document to a Document
```
Nitpick: I think `an` is actually correct here 😀

malte-laukoetter replied:
atleast intellij complaint about "an" but I don't know 🤷

andreasphil commented on frontend/src/components/references/RisModSelectionPanel.vue:
```diff
@@ -0,0 +1,57 @@
+import RisCallout from "@/components/controls/RisCallout.vue"
+
+const props = defineProps<{
+  normXml: string
```
Todo: docs

andreasphil commented on frontend/src/components/references/RisRefSelectionPanel.vue:
```diff
@@ -0,0 +1,140 @@
+</template>
+
+<style scoped>
+:deep(.akn-quotedStructure .akn-quote-startQuote) {
```
Same here as above regarding the `deep` selectors

andreasphil commented on frontend/src/composables/useAknTextSelection.ts:
```diff
@@ -0,0 +1,106 @@
+  }
+}
+
+export function useAknTextSelection(
```
todo: docs

andreasphil commented on frontend/src/lib/htmlRangeToLdmlDeRange.ts:
```diff
@@ -0,0 +1,129 @@
+}
+
+/**
+ * Finds the offset that is the corresponds to the same text range in the LDML.de node.
```
todo: IMO this could need some additional explanation about the algorithm that is used for finding the offset

andreasphil commented on frontend/src/lib/htmlRangeToLdmlDeRange.ts:
```diff
@@ -0,0 +1,129 @@
+}
+
+/**
+ * Converts a Range within a html-render of a LMDL.de xml to a range within that xml document.
```
todo: same here, some explanation about the algorithm would be appreciated

andreasphil commented on frontend/src/lib/ref.ts:
```diff
@@ -0,0 +1,50 @@
+    "ref",
+  )
+  aknRef.setAttribute("eId", getNextRefEId(parentElement))
+  aknRef.setAttribute("GUID", uuidv4())
```
Question: is this a temporary GUID or will this also be the GUID in the backend? If it's the latter, it would feel better to me to have one actor in charge of generating IDs (in our case: the backend) to ensure data integrity and uniqueness (even though I admit that the likelihood of conflicts is low).

malte-laukoetter replied:
It's the final one. But as we are using UUIDs here (and especially v4 once) we do not have any data integrity or uniqueness problems as they are designed with the explicit purpose that lots of distributed actors can generate them and there will not be any conflicts. So it is totally fine to create the GUID on the frontend.

andreasphil commented on frontend/src/views/References.vue:
```diff
@@ -0,0 +1,221 @@
+
+    parentNode.replaceChild(newNode, selectedModQuotedContent.value)
+    // we mutate the norm Document (as a side effect), so we need to trigger a recompute of all things that use it
+    triggerRef(amendingNormDocument)
```
I didn't know about `triggerRef`, nice! 💡

malte-laukoetter committed:
Improve docs

RISDEV-4397

malte-laukoetter committed:
Improve loading indicator

Previously it was shown quite often when marking new akn:ref elements
and the preview was always hidden. This felt a bit annoying.

RISDEV-4397

malte-laukoetter committed:
Fix tests

RISDEV-4397
